File: public/autocomplete.js
// Copyright © 2014-2017 Lennart Bierkandt <post@lennartbierkandt.de>
//
// This file is part of GraphAnno.
//
// GraphAnno is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// GraphAnno is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with GraphAnno. If not, see <http://www.gnu.org/licenses/>.

var Autocomplete = (function(){
	var $element = null;
	var $list = null;
	var value = '';
	var noInput = false;

	var parseInput = function() {
		var cursorPosition = $element[0].selectionDirection == 'backward' ? $element[0].selectionStart : $element[0].selectionEnd;
		var string = $element.val();
		var command = string.match(/^\s*\S+\s/) ? string.match(/^\s*(\S+)/)[1] : '';
		var suggestionSet = window.autocompleteCommands[command];
		var before = string.slice(0, cursorPosition).match(suggestionSet == 'file' ? /^(.*(\s|^))(\S*)$/ : /^(.*(\.\.|\s|^))(\S*)$/);
		var after = string.slice(cursorPosition).match(/^\s*(.*)$/);
		var word = string.slice(cursorPosition).match(/^(\s|$)/) ? before[3] : '';
		var sep = suggestionSet == 'file' ? '/' : null;
		var wordParts = word.match(sep) ? word.match('^(.*' + sep + ')([^'+ sep + ']*)$') : [null, '', word];
		return {
			before: before[1],
			command: command,
			word: word,
			retain: wordParts[1],
			replace: wordParts[2],
			after: after[1],
			suggestionSet: suggestionSet,
			layer: document.cmd.layer.value,
		};
	}
	var setList = function(words) {
		$list.html('').scrollTop(0);
		words.forEach(function(word){
			$list.append('<div>' + word + '</div>');
		});
		$list.children(':first-child').addClass('active');
	}
	var insert = function() {
		var word = $list.find('.active').text();
		var input = parseInput();
		if (input.suggestionSet == 'file' && word.match(/\/$/)) var upToCursor = input.before + input.retain + word;
		else var upToCursor = input.before + input.retain + word + ' ';
		$element.val(upToCursor + input.after);
		$element[0].setSelectionRange(upToCursor.length, upToCursor.length);
		if (input.suggestionSet == 'command') handleInput();
		if (input.suggestionSet == 'file' && word.match(/\/$/)) handleInput();
	}
	var disable = function() {
		$list.hide();
		$element.off('.autocomplete');
	}
	var changeActive = function(keycode) {
		var active = $list.find('.active');
		if (keycode == 38) {
			var newActive = active.prev();
			if (newActive.length == 0) newActive = $list.children(':last-child');
		} else if (keycode == 40) {
			var newActive = active.next();
			if (newActive.length == 0) newActive = $list.children(':first-child');
		}
		$list.children().removeClass('active');
		newActive.addClass('active');
		scrollIntoView('.active');
	}
	var scrollIntoView = function () {
		var $element = $('#autocomplete .active');
		if ($element.length == 0) return;
		var $container = $('#autocomplete');
		var containerViewTop = $container.scrollTop();
		var containerViewHeight = $container[0].clientHeight;
		var containerViewBottom = containerViewTop + containerViewHeight;
		var elementTop = Math.ceil($element.position().top + containerViewTop);
		var elementBottom = Math.ceil(elementTop + $element.height());
		if (elementTop < containerViewTop)
			$container.scrollTop($element.position().top);
		else if (elementBottom > containerViewBottom) {
			$container.scrollTop(elementTop + $element.height() - containerViewHeight);
		}
	}
	var keyBinding = function(e) {
		var actions = {
			9: function() {insert(); disable();},
			27: disable,
			38: changeActive,
			40: changeActive,
		};
		if (e.which in actions) {
			e.preventDefault();
			noInput = true;
			actions[e.which](e.which);
		}
	}
	var saveValue = function() {
		value = $element.val();
	}
	var valueUnchanged = function() {
		return $element.val() == value;
	}
	var handleInput = function(e) {
		if (!window.preferences.autocompletion) return;
		if (valueUnchanged()) return;
		if (noInput) {noInput = false; return;}
		var input = parseInput();
		if (input.word.length > 0 || input.suggestionSet == 'file') {
			if (!input.suggestionSet) return;
			$.getJSON('/get_autocomplete_suggestions/', input).done(function(suggestions){
				showSuggestions(input, suggestions);
			});
		} else {
			disable();
		}
	}
	var showSuggestions = function(input, suggestions) {
		if (suggestions.length > 0) {
			setList(suggestions.sort());
			if ($list.css('display') == 'none') {
				var coordinates = getCaretCoordinates($element[0], $element[0].selectionEnd);
				$list.css({left: coordinates.left}).show();
				$element.on('keydown.autocomplete', keyBinding);
			}
		} else {
			disable();
		}
	}

	return {
		init: function(selector) {
			$element = $(selector);
			$list = $('<div id="autocomplete"></div>').appendTo($element.parent());
			$element.on('keydown', saveValue);
			$element.on('keyup', handleInput);
		},
		disable: disable,
	}
})();

File: lib/controller/autocomplete.rb
# encoding: utf-8

# Copyright © 2014-2017 Lennart Bierkandt <post@lennartbierkandt.de>
#
# This file is part of GraphAnno.
#
# GraphAnno is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# GraphAnno is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with GraphAnno. If not, see <http://www.gnu.org/licenses/>.

module Autocomplete
	def get_autocomplete_suggestions
		input = @sinatra.params[:word]
		data = case @sinatra.params[:suggestionSet].to_sym
		when :file
			return file_suggestions(input).to_json
		when :anno
			tagset_suggestions(@sinatra.params) + makro_suggestions + aref_suggestions
		when :nodes_anno
			tagset_suggestions(@sinatra.params) + makro_suggestions + layer_suggestions +
				nnode_suggestions + inode_suggestions + token_suggestions
		when :edges_anno
			tagset_suggestions(@sinatra.params) + makro_suggestions + layer_suggestions + edge_suggestions
		when :nodes
			nnode_suggestions + inode_suggestions
		when :nnodes
			nnode_suggestions
		when :inodes
			inode_suggestions
		when :tokens
			token_suggestions
		when :ref
			ref_suggestions
		when :layer
			layer_suggestions + ref_suggestions
		when :sect
			section_suggestions
		when :annotator
			annotator_suggestions
		when :command
			command_suggestions
		end
		data.select do |suggestion|
			suggestion.to_s.start_with?(input)
		end.to_json
	end

	private

	def self.commands
		{
			:a => :anno,
			:n => :nodes_anno,
			:e => :nodes_anno,
			:p => :nodes_anno,
			:g => :nodes_anno,
			:c => :nodes_anno,
			:h => :nodes_anno,
			:ni => :edges_anno,
			:di => :nodes,
			:do => :nodes,
			:sa => :inodes,
			:sd => :nnodes,
			:d => :ref,
			:t => nil,
			:tb => nil,
			:ta => nil,
			:undo => nil,
			:z => nil,
			:redo => nil,
			:y => nil,
			:l => :layer,
			:annotator => :annotator,
			:user => :annotator,
			:ns => nil,
			:'s-new' => :sect,
			:'s-rem' => :sect,
			:'s-add' => :sect,
			:'s-det' => :sect,
			:'s-del' => :sect,
			:load => :file,
			:add => :file,
			:append => :file,
			:save => nil,
			:clear => nil,
			:s => :sect,
			:image => nil,
			:export => nil,
			:import => nil,
			:play => :tokens,
			:config => nil,
			:tagset => nil,
			:makros => nil,
			:metadata => nil,
			:annotators => nil,
			:file => nil,
			:pref => nil,
			:'' => :command
		}
	end

	def tagset_suggestions(params)
		parameters = params[:before].parse_parameters
		if params[:suggestionSet] == 'anno'
			@graph.tagset.for_autocomplete(
				extract_elements(parameters[:elements])
			)
		else
			layers = if layer_shortcut = get_layer_shortcut(parameters[:words])
				@graph.conf.layer_by_shortcut[layer_shortcut]
			else
				@graph.conf.layer_by_shortcut[params[:layer]]
			end
			element = if params[:command] == 'e'
				@graph.create_phantom_edge(:type => 'a', :layers => layers)
			else
				@graph.create_phantom_node(:type => 'a', :layers => layers)
			end
			@graph.tagset.for_autocomplete([element])
		end
	end

	def file_suggestions(input)
		relative = input[0] != '/'
		Dir.glob("#{'data/' if relative}#{input}*").map{|file|
			if File.directory?(file)
				file.sub!(/^data\//, '') if relative
				# strip path and add trailing slash
				file.sub(/^.*\/([^\/]+)$/, '\1') + '/'
			else
				# exclude non-json and log files, strip path
				(file.match(/\.json$/) && !file.match(/log\.json$/)) ? file.sub(/^(.+\/)?([^\/]+)$/, '\2') : nil
			end
		}.compact
	end

	def makro_suggestions
		@preferences[:makro] ? @graph.anno_makros.keys : []
	end

	def layer_suggestions
		@preferences[:makro] ? @graph.conf.layer_by_shortcut.keys : []
	end

	def token_suggestions
		@preferences[:ref] ? @view.tokens.map.with_index{|t, i| "t#{i}"} : []
	end

	def nnode_suggestions
		@preferences[:ref] ? @view.dependent_nodes.map.with_index{|n, i| "n#{i}"} : []
	end

	def inode_suggestions
		@preferences[:ref] ? @view.i_nodes.map.with_index{|n, i| "i#{i}"} : []
	end

	def edge_suggestions
		@preferences[:ref] ? @view.edges.map.with_index{|e, i| "e#{i}"} : []
	end

	def ref_suggestions
		token_suggestions + nnode_suggestions + inode_suggestions + edge_suggestions
	end

	def sref_suggestions
		(sections = @graph.sections_hierarchy(@current_sections)) ? sections.map.with_index{|s, i| "s#{i}"} : []
	end

	def aref_suggestions
		@preferences[:ref] ? ref_suggestions + sref_suggestions : []
	end

	def section_suggestions
		@preferences[:sect] ? @graph.section_nodes.map(&:name).compact : []
	end

	def annotator_suggestions
		@graph.annotators.map(&:name)
	end

	def command_suggestions
		@preferences[:command] ? Autocomplete.commands.keys : []
	end
end
